### PR TITLE
Add array_has_duplicates Presto function

### PR DESCRIPTION
**Pull request by duanmeng**

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Any particular reason to implement this function as a vector function? It seems it can be implemented as a simple function, no?

**Comment by duanmeng** [comment]
> Details
@mbasmanova  I follow the guidance of developer guide, which mentioned "[Simple functions process a single row and produce a single value as a result. Vector functions process a batch or rows and produce a vector of results. ](https://facebookincubator.github.io/velox/develop/scalar-functions.html#vector-functions:~:text=Simple%20functions%20process%20a%20single%20row%20and%20produce%20a%20single%20value%20as%20a%20result.%20Vector%20functions%20process%20a%20batch%20or%20rows%20and%20produce%20a%20vector%20of%20results.)". And I reference the implementation of ArrayDuplicates and ArrayContains.

**Comment by mbasmanova** [comment]
@duanmeng I suggest to implement this function as a simple function. (ArrayDuplicates and ArrayContains may have been implemented before simple functions supported complex types. It would be nice to convert these to simple functions as well.)

**Comment by duanmeng** [comment]
> @duanmeng I suggest to implement this function as a simple function. (ArrayDuplicates and ArrayContains may have been implemented before simple functions supported complex types. It would be nice to convert these to simple functions as well.)

@mbasmanova Got it, I will change my implementation.

**Comment by duanmeng** [comment]
@mbasmanova I've updated the implementation using simple functions, could you please help to review?

File: velox/functions/prestosql/ArrayFunctions.h [hunk 1]
```diff
@@ -316,4 +316,33 @@ struct ArraySumFunction {
...
+  FOLLY_ALWAYS_INLINE void call(
+      bool& out,
+      const arg_type<velox::Array<T>>& inputArray) {
+    folly::F14FastSet<arg_type<T>> uniqSet;
+    int16_t numNulls = 0;
+    out = false;
+    for (const auto& item : inputArray) {
+      if (item.has_value()) {
+        if (uniqSet.find(item.value()) == uniqSet.end()) {
+          uniqSet.insert(item.value());
```
Comment (mbasmanova) [inline velox/functions/prestosql/ArrayFunctions.h:332]: You can call insert unconditionally and check the result to determine if the item is a duplicate or not.

```
if (!uniqSet.insert(item.value().second) {
   // Found duplicate.
}
```

File: velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp [hunk 2]
```diff
@@ -50,6 +50,14 @@ inline void registerArrayCombinationsFunctions() {
       int64_t>({"combinations"});
 }
 
+template <typename T>
+inline void registerArrayHasDuplicationsFunctions() {
```
Comment (mbasmanova) [inline velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp:54]: registerArrayHasDuplicationsFunctions -> ...Duplicates...

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+#include <optional>
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::functions::test;
+
+namespace {
+
+// Class to test the array_has_duplicates operator.
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:26]: This comment is redundant. Let's remove.

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+    testExpr(expected, "array_has_duplicates(C0)", {array});
+  }
+};
+
+} // namespace
+
+// Test integer arrays.
+TEST_F(ArrayHasDuplicatesTest, integerArrays) {
+  testBigint();
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:74]: This method is used only once. Consider removing it and adding code here directly.

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        {false, false, false, false, true, true, true, true, true, true});
+
+    testExpr(expected, "array_has_duplicates(C0)", {array});
+  }
+};
+
+} // namespace
+
+// Test integer arrays.
+TEST_F(ArrayHasDuplicatesTest, integerArrays) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:73]: integerArrays -> bigints or bigint

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+  testBigint();
+}
+
+// Test inline (short) strings.
+TEST_F(ArrayHasDuplicatesTest, inlineStringArrays) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:78]: inlineStringArrays -> inlineStrings

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+  using S = StringView;
+
+  auto array = makeNullableArrayVector<StringView>({
+      {},
+      {S("")},
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:83]: use ""_sv instead

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+  });
+
+  auto expected = makeFlatVector<bool>(
+      {false, false, false, false, false, true, true, true, true});
+
+  testExpr(expected, "array_has_duplicates(C0)", {array});
+}
+
+// Test non-inline (> 12 character length) strings.
+TEST_F(ArrayHasDuplicatesTest, stringArrays) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:100]: stringArrays -> longStrings

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+          std::nullopt,
+          S("purple is an elegant color"),
+      },
+  });
+
+  auto expected = makeFlatVector<bool>({false, true, true});
+  testExpr(expected, "array_has_duplicates(C0)", {array});
+}
+
+TEST_F(ArrayHasDuplicatesTest, nonContiguousRows) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:126]: This test is redundant as this logic is processed by the framework.

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+  auto expected = makeFlatVector<bool>({true, false, true, false});
+
+  auto result = evaluate<SimpleVector<bool>>(
+      "if(c0 % 2 = 0, array_has_duplicates(c1), array_has_duplicates(c2))",
+      makeRowVector({c0, c1, c2}));
+  assertEqualVectors(expected, result);
+}
+
+TEST_F(ArrayHasDuplicatesTest, constant) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:150]: same

File: velox/functions/prestosql/tests/CMakeLists.txt [hunk 11]
```diff
@@ -25,6 +25,7 @@ add_executable(
   ArrayExceptTest.cpp
   ArrayFilterTest.cpp
   ArrayJoinTest.cpp
+  ArrayHasDuplicatesTest
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/CMakeLists.txt:28]: .cpp is missing

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Thank you for iterating on this PR. Overall looks good.

Please, update the PR title to use the template from https://github.com/facebookincubator/velox/blob/main/CONTRIBUTING.md#prestos-sql-functions and add documentation.

Please, also run Fuzzer:

```
velox_expression_fuzzer_test --logtostderr=1 --velox_fuzzer_enable_complex_types --only array_has_duplicates --duration_sec 60

velox_expression_fuzzer_test --logtostderr=1 --velox_fuzzer_enable_complex_types --enable_variadic_signatures --duration_sec 600
```

File: velox/functions/prestosql/ArrayFunctions.h (hunk 1 above)
Comment (duanmeng) [reply velox/functions/prestosql/ArrayFunctions.h]: Fixed

File: velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp (hunk 2 above)
Comment (duanmeng) [reply velox/functions/prestosql/registration/ArrayFunctionsRegistration.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 3 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 4 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 5 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 6 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 7 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 8 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 9 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Fixed

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 10 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Fixed

File: velox/functions/prestosql/tests/CMakeLists.txt (hunk 11 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/CMakeLists.txt]: Fixed

**Comment by duanmeng** [comment]
@mbasmanova  I've resolved the comments and updated the PR title, and doc, and ran the two fuzzer tests following your guidance.

The partial log of the two tests
```
./_build/release/velox/expression/tests/velox_expression_fuzzer_test --logtostderr=1 --velox_fuzzer_enable_complex_types --only array_has_duplicates --duration_sec 60
....
I20221122 23:23:34.645284 11134596 ExpressionVerifier.cpp:129] Executing expression: array_has_duplicates(10 elements starting at 630 {[630->629] 1495423791, [631->264] 555425632, [632->211] 1155354488, [633->22] 1768542804, [634->245] 1479300114, ...})
I20221122 23:23:34.645294 11134596 ExpressionVerifier.cpp:42] 0 vectors as input:
I20221122 23:23:34.645351 11134596 ExpressionVerifier.cpp:79] All results match.
I20221122 23:23:34.645359 11134596 ExpressionFuzzer.cpp:704] ==============================> Done with iteration 472691
I20221122 23:23:34.645367 11134596 ExpressionFuzzer.cpp:656] ==============================> Started iteration 472692 (seed: 4133578585)
I20221122 23:23:34.645371 11134596 ExpressionFuzzer.cpp:540] Couldn't find any function to return 'ARRAY<VARCHAR>'. Returning a constant instead.
I20221122 23:23:34.645615 11134596 ExpressionVerifier.cpp:129] Executing expression: array_has_duplicates(10 elements starting at 320 {Q.ZA&G3``rBE?)dP!W}+z, gPJ).q]:gio!sY~a-BRBj>6PA%Mg_oeI^9E, h'*BcJ'.:B,\Mq".~ibXcl9Brh\5)]sI^O"g6:I%K(U#*c[, eJMJ0dwF2L3D0l^N,6p-<`e*n!N,Kp<]NG4,u?s\yIp>bSaz-6A&CdY+, .#g5@>ac"mK#XKl[f;cSao$|t%<j#?w2W&4}Sv]v"(^QJ-V&ue^P{o}q4rmDx9xGBlbXYdrY6N#, ...})
I20221122 23:23:34.645627 11134596 ExpressionVerifier.cpp:42] 0 vectors as input:
I20221122 23:23:34.645705 11134596 ExpressionVerifier.cpp:79] All results match.
I20221122 23:23:34.645711 11134596 ExpressionFuzzer.cpp:704] ==============================> Done with iteration 472692
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
[macduan@MACDUAN-MB0~/work/native/velox]
$ echo $?
0

./_build/release/velox/expression/tests/velox_expression_fuzzer_test --logtostderr=1 --velox_fuzzer_enable_complex_types --enable_variadic_signatures --duration_sec 600
...
E20221122 23:21:51.354403 11126525 Exceptions.h:68] Line: /Users/macduan/work/native/velox/velox/functions/prestosql/Map.cpp:101, Function:operator(), Expression: !keysElements->isNullAt(offset + i) map key cannot be null, Source: USER, ErrorCode: INVALID_ARGUMENT
E20221122 23:21:51.360077 11126525 Exceptions.h:68] Line: /Users/macduan/work/native/velox/velox/functions/lib/DateTimeFormatter.cpp:1164, Function:buildMysqlDateTimeFormatter, Expression:  Both printing and parsing not supported, Source: USER, ErrorCode: INVALID_ARGUMENT
E20221122 23:21:51.363798 11126525 Exceptions.h:68] Line: /Users/macduan/work/native/velox/velox/functions/prestosql/Map.cpp:101, Function:operator(), Expression: !keysElements->isNullAt(offset + i) map key cannot be null, Source: USER, ErrorCode: INVALID_ARGUMENT
I20221122 23:21:51.364099 11126525 ExpressionFuzzer.cpp:704] ==============================> Done with iteration 360421
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
[macduan@MACDUAN-MB0~/work/native/velox]
$ echo $?
0
```

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp (hunk 7 above)
Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp]: Should I just fix this empty string "", or replace all the S("str") with "str"_sv?

**Comment by duanmeng** [comment]
@mbasmanova Could you please help to take a look?

**Comment by duanmeng** [comment]
@xiaoxmeng @Yuhta Could you help to review this pr, I've resolved all the comments from @mbasmanova, she seems to be on vacation.

**Review by mbasmanova — APPROVED** [review]
@duanmeng Looks good to me. Thank you for the contribution.

**Comment by duanmeng** [comment]
@mbasmanova  Thanks for your help on my first add function pr in velox:). Can I do something about the *Facebook internal builds & tests*?

File: velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+       {S("red shiny car ahead"),
+        S("blue clear sky above"),
+        S("blue clear sky above")},
+       {S("a"), S("b")},
+       {S("a"), S("b"), S("b")}
+
+      });
+  auto expected = makeFlatVector<bool>({false, true, false, true});
+  testExpr(expected, "array_has_duplicates(C0)", {array});
+}
```
Comment (darrenfu) [inline velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:147]: Thanks for the contribution, @duanmeng. I have closed my duplicate PR. Feel free to borrow more unit test cases, such as constant vectors, nonContiguousRows, here.

https://github.com/facebookincubator/velox/pull/3397/files#diff-198e97a86767f3f7b75862d0350895b93496dc376dbf8a1fe3bc384cec368076

Comment (duanmeng) [reply velox/functions/prestosql/tests/ArrayHasDuplicatesTest.cpp:147]: > Thanks for the contribution, @duanmeng. I have closed my duplicate PR. Feel free to borrow more unit test cases, such as constant vectors, nonContiguousRows, here.
> 
> https://github.com/facebookincubator/velox/pull/3397/files#diff-198e97a86767f3f7b75862d0350895b93496dc376dbf8a1fe3bc384cec368076

@darrenfu Thanks for your suggestion, I used to add the constant and non Contiguous Rows test case referring to the test case in `Array Duplicates Test` and removed them according to the comments *"This test is redundant as this logic is processed by the framework"*.

**Comment by mbasmanova** [comment]
@duanmeng Thank you for the contribution.

**Pull request closed without merging**